src/encoding.rs

```diff
@@ -690,75 +689,76 @@ macro_rules! length_delimited {
                     .map(|value| encoded_len_varint(value.len() as u64) + value.len())
                     .sum::<usize>()
         }
-
-        #[cfg(test)]
-        mod test {
-            use quickcheck::{quickcheck, TestResult};
-
-            use super::super::test::{check_collection_type, check_type};
-            use super::*;
-
-            quickcheck! {
-                fn check(value: $ty, tag: u32) -> TestResult {
-                    super::test::check_type(value, tag, WireType::LengthDelimited,
-                                            encode, merge, encoded_len)
-                }
-                fn check_repeated(value: Vec<$ty>, tag: u32) -> TestResult {
-                    super::test::check_collection_type(value, tag, WireType::LengthDelimited,
-                                                       encode_repeated, merge_repeated,
-                                                       encoded_len_repeated)
-                }
-            }
-        }
     };
 }
 
 pub mod string {
     use super::*;
 
-    pub fn encode<B>(tag: u32, value: &String, buf: &mut B)
+    pub fn encode<B>(tag: u32, value: &BytesString, buf: &mut B)
     where
         B: BufMut,
     {
         encode_key(tag, WireType::LengthDelimited, buf);
         encode_varint(value.len() as u64, buf);
-        buf.put_slice(value.as_bytes());
+        buf.put_slice(value.into());
     }
-    pub fn merge<B>(wire_type: WireType, value: &mut String, buf: &mut B) -> Result<(), DecodeError>
+    pub fn merge<B>(
+        wire_type: WireType,
+        value: &mut BytesString,
+        buf: &mut B,
+    ) -> Result<(), DecodeError>
     where
         B: Buf,
     {
         unsafe {
-            // String::as_mut_vec is unsafe because it doesn't check that the bytes
+            // BytesString::as_bytes_mut is unsafe because it doesn't check that the bytes
             // inserted into it the resulting vec are valid UTF-8. We check
             // explicitly in order to ensure this is safe.
-            super::bytes::merge(wire_type, value.as_mut_vec(), buf)?;
+
+            super::bytes::merge(wire_type, value.as_bytes_mut(), buf)?;
             str::from_utf8(value.as_bytes())
                 .map_err(|_| DecodeError::new("invalid string value: data is not UTF-8 encoded"))?;
```

Comment:
This looks wrong: `value` is left with broken UTF-8. The previous code was wrong too, FWIW.

`String::as_mut_vec` should be banned and replaced with a more principled scope guard API that would sanitize the buffer on unwrap. I've been too lazy to write an RFC on this.

Reply:
@mzabaluev that's a great catch.  Any interest in sending a PR?  I imagine the fix will need to be something along the lines of swapping the string field with an empty string (to retain any allocated capacity), converting to a `Vec<u8>`, copy the bytes in, then converting back to a String and storing it?

IIUC the only way this could have been caught without a specific unit test is if we ran the fuzzer under MIRI, but I'm not sure if that's technically possible.  In any case, it should be straightforward to check with a unit test.

Reply:
It'd be worthwhile to audit for any other uses of `as_mut_vec()` as well, you are definitely right that that's a footgun of an API.

Reply:
#194 has a fix for the master branch. That's the only use of `String::as_mut_vec` in the repository.

Reply:
Just to make sure I understand the problem here - the issue is if there is an error in `merge` and we implicitly return via the `?`. Do we assume that `value` is valid in the error case? I suppose the user is able to do whatever they want to so we shouldn't give them invalid data.

Reply:
@nrc It is UB to expose invalid UTF-8 data inside a `String`. Both the error case, and any panic in the `Buf` implementation, may result in a malformed string being accessed.